### PR TITLE
env::temp_dir returns /private/tmp on Apple instead while /tmp is

### PR DESCRIPTION
**Pull request by devnexen**
a symlink in fact.

ref #99608

### CODE REVIEW

**Comment by thomcc** [comment]
Hmm, I am not so sure about this.

The [preferred](https://developer.apple.com/library/archive/documentation/Security/Conceptual/SecureCodingGuide/Articles/RaceConditions.html#//apple_ref/doc/uid/TP40002585-SW10) tempdir on these platforms is the value from `libc::confstr(libc::_CS_DARWIN_USER_TEMP_DIR, ...)`, although this will be a user-specific directory with a name like `/var/folders/rw/gsmfnk2s12d34hywmktc28q40000gn/T/`, where the specific name will change over time and is unpredictable.

This is best for security, as it avoids providing a temporary directory that another user (who may be at a different privilege level) has access to, unlike /tmp or /private/tmp. Usually `$TMPDIR` is set to this, but I guess not always? That said, I don't know if using this is a change we'd want to make.

More generally, we've always returned `/tmp` on darwin platforms, despite it (AFAIK) always being a symlink to `/private/tmp`. So I don't really think changing it to this is what I'd recommend, If we're going to change it, I'd rather see us use the safer directory.

Either way, this is a behavioral change that would require [documentation](https://doc.rust-lang.org/nightly/std/env/fn.temp_dir.html#platform-specific-behavior) to be updated, and t-libs-api signoff.

@rustbot label +T-libs-api -T-libs

**Comment by thomcc** [comment]
I think this needs further discussion in the t-libs-api meeting. Essentially there are three choices here for what to do if `$TMPDIR` is unset on Darwin platforms:

1. Keep the same behavior we have currently, which is to return `/tmp`, which is (and has always been) a symlink to `/private/tmp`.

2. Change to returning `/private/tmp`. This avoids returning something that is symlink, and returns what that link resolves to. The rationale here is certain programs may not expect the path returned to be a symlink (although I'm not totally certain that it's never a 

3. Change to returning the string . This matches the behavior used when `$TMPDIR

**Comment by thomcc** [comment]
(Sorry for the accidental close -- accidentally hit tab). See previous message for explanation.

@rustbot label +I-libs-api-nominated

**Comment by m-ou-se** [comment]
> The rationale here is certain programs may not expect the path returned to be a symlink

Are there any examples of this? Is there any realistic use case where a symlink would be a problem?

**Comment by thomcc** [comment]
It did cause an issue for someone [using SQLite](https://github.com/rusqlite/rusqlite/issues/1206#issuecomment-1192635895), since they were requesting SQLite not follow symbolic links.

That said, I don't think most software on Darwin OSes should be using `/tmp` *or* `/private/tmp` in most cases. Doing so goes against [Apple's security guidelines](https://developer.apple.com/library/archive/documentation/Security/Conceptual/SecureCodingGuide/Articles/RaceConditions.html#//apple_ref/doc/uid/TP40002585-SW10), which tell you to use `confstr(_CS_DARWIN_USER_TEMP_DIR, ...)` instead.

Generally this is not needed, as most of the time `$TMPDIR` is set in the environment to the directory the `confstr` invocation will resolve to. I guess in some cases it is not though, and think that we should use that as a fallback instead if we can justify the behavioral change.

File: library/std/src/sys/unix/os.rs [hunk 1]
```diff
@@ -586,6 +586,36 @@ pub fn temp_dir() -> PathBuf {
...
             PathBuf::from("/data/local/tmp")
+        } else if cfg!(target_vendor = "apple") {
+            extern "C" {
+                fn confstr(
+                    name: libc::c_int,
+                    buf: *mut libc::c_char,
+                    len: libc::size_t,
+                ) -> libc::size_t;
+            }
+            let tmpdir = unsafe { libc::getenv(b"TMPDIR".as_ptr() as *const libc::c_char) };
```
Comment (thomcc) [inline library/std/src/sys/unix/os.rs:597]: We already checked the env above. Also, don't use `libc::getenv` as it doesn't hold the env lock. Also, this string isn't NUL-terminated.

File: library/std/src/sys/unix/os.rs [hunk 2]
```diff
@@ -586,6 +586,36 @@ pub fn temp_dir() -> PathBuf {
...
+            if tmpdir.is_null() {
+                let mut buf: Vec<u8> = vec![0; libc::PATH_MAX as usize];
+                const _CS_DARWIN_USER_TEMP_DIR: libc::c_int = 65537;
```
Comment (thomcc) [inline library/std/src/sys/unix/os.rs:600]: Let's wait for https://github.com/rust-lang/libc/pull/2883.

File: library/std/src/sys/unix/os.rs [hunk 3]
```diff
@@ -586,6 +586,36 @@ pub fn temp_dir() -> PathBuf {
...
+            extern "C" {
+                fn confstr(
+                    name: libc::c_int,
+                    buf: *mut libc::c_char,
+                    len: libc::size_t,
+                ) -> libc::size_t;
+            }
+            let tmpdir = unsafe { libc::getenv(b"TMPDIR".as_ptr() as *const libc::c_char) };
+            if tmpdir.is_null() {
+                let mut buf: Vec<u8> = vec![0; libc::PATH_MAX as usize];
```
Comment (thomcc) [inline library/std/src/sys/unix/os.rs:599]: 50 bytes should be enough in practice, but really this should have some retry logic. I was going to submit https://github.com/thomcc/rust/commit/ee1c648d4be0b9c23399450578d434018db8868b as a PR at one point, but you can just integrate the changes from it.

**Review by thomcc — CHANGES_REQUESTED** [review]

File: library/std/src/sys/unix/os.rs (hunk 3 above)
Comment (devnexen) [reply library/std/src/sys/unix/os.rs:599]: it s better if you do since you authored all.

**Pull request closed without merging**